File: Python/ps_baekjoon/14719/main.py
import sys
from queue import LifoQueue

readline, write = (sys.stdin.readline, sys.stdout.write)
s = LifoQueue()
res = 0

h, w = [int(i) for i in readline().split()]
board = [[False] * (w + 1) for _ in range(h + 1)]
block = [None] + [int(i) for i in readline().split()]
for col in range(1, w + 1):
    for row in range(h - block[col] + 1, h + 1):
        board[row][col] = True
for row in range(h, 0, -1):
    s.queue.clear()
    for col in range(1, w + 1):
        if board[row][col]:
            if not s.empty():
                res += (col - s.get()) - 1
            s.put(col)
write(f"{res}")

File: C++/ps_baekjoon/14719/main.cpp
#include <bits/stdc++.h>
using namespace std;

constexpr int mxH = 5e2, mxW = 5e2;
int h, w, res = 0, block[mxW + 1];
bool board[mxH + 1][mxW + 1];
stack<int> s;

int main()
{
    cin.tie(0)->sync_with_stdio(0);
    cin >> h >> w;
    for (int i = 1; i <= w; ++i)
        cin >> block[i];
    for (int col = 1; col <= w; ++col)
        for (int row = h - block[col] + 1; row <= h; ++row)
            board[row][col] = true;
    for (int row = h; row >= 1; --row)
    {
        while (!s.empty())
            s.pop();
        for (int col = 1; col <= w; ++col)
            if (board[row][col])
            {
                if (!s.empty())
                {
                    res += col - s.top() - 1;
                    s.pop();
                }
                s.push(col);
            }
    }
    cout << res;
    return 0;
}